### PR TITLE
www-apps/redmine: updates

### PR DESCRIPTION
**Pull request by winterheart**
Updated to new supported versions 3.4.11, 4.0.4, removed old.

Fixed bugs:
* https://bugs.gentoo.org/677986
* https://bugs.gentoo.org/692890
* https://bugs.gentoo.org/657156
* https://bugs.gentoo.org/660252

### CODE REVIEW

File: www-apps/redmine/redmine-3.4.11.ebuild [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+	>=dev-ruby/roadie-rails-1.1.1
+	>=dev-ruby/rbpdf-1.19.6
+	>=dev-ruby/ruby-openid-2.3.0
+	dev-ruby/rubygems
+	"
+
+REDMINE_DIR="/var/lib/${PN}"
+
+pkg_setup() {
+	enewgroup redmine
```
Comment (juippis) [inline www-apps/redmine/redmine-3.4.11.ebuild]: Needs to be handled via the GLEP-81 way, many examples already on gentoo-dev mailing list.

File: www-apps/redmine/redmine-3.4.11.ebuild [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+	dev-ruby/protected_attributes
+	>=dev-ruby/rack-openid-0.2.1
+	>=dev-ruby/rails-4.2.11.1:4.2
+	>=dev-ruby/rails-html-sanitizer-1.0.3
+	dev-ruby/request_store:1.0.5
+	>=dev-ruby/roadie-rails-1.1.1
+	>=dev-ruby/rbpdf-1.19.6
+	>=dev-ruby/ruby-openid-2.3.0
+	dev-ruby/rubygems
+	"
```
Comment (juippis) [inline www-apps/redmine/redmine-3.4.11.ebuild]: This tab is useless, isn't it?

**Comment by juippis** [comment]
Also, 
```
RepoMan-Options: --force
```
why?

**Comment by winterheart** [comment]
@juippis updated to GLEP81, taken UID/GUID 451

**Comment by juippis** [comment]
Hmm, could you clean and re-order this PR? For example, these bumps to 3.4.11 and 4.0.4 are pretty useless since no one is ever going to install them. Then you could switch to acct- method on version bump, not in a separate "fixing" commit.

**Comment by winterheart** [comment]
@juippis done.

File: acct-group/redmine/metadata.xml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
+<maintainer type="person">
```
Comment (juippis) [inline acct-group/redmine/metadata.xml]: Please, remember proper indentation here.

File: acct-user/redmine/metadata.xml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
+<maintainer type="person">
```
Comment (juippis) [inline acct-user/redmine/metadata.xml]: And here.

File: www-apps/redmine/redmine-3.4.12.ebuild [hunk 5]
```diff
@@ -0,0 +1,232 @@
+# Copyright 1999-2019 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=6
```
Comment (juippis) [inline www-apps/redmine/redmine-3.4.12.ebuild]: One empty line below EAPI please. Also please split updating these files into their own commits.

File: www-apps/redmine/redmine-3.4.12.ebuild [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+USE_RUBY="ruby23 ruby24"
```
Comment (juippis) [inline www-apps/redmine/redmine-3.4.12.ebuild]: ruby-2.3 has been removed from Gentoo, could you test this with 2.5 and 2.6 instead?

File: www-apps/redmine/redmine-3.4.12.ebuild [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+inherit eutils depend.apache ruby-ng
+
+DESCRIPTION="Flexible project management web application using the Ruby on Rails framework"
+HOMEPAGE="https://www.redmine.org/"
+SRC_URI="https://www.redmine.org/releases/${P}.tar.gz"
+
+KEYWORDS="~amd64"
+LICENSE="GPL-2"
+SLOT="0"
+IUSE="imagemagick fastcgi ldap markdown mysql passenger postgres sqlite"
```
Comment (juippis) [inline www-apps/redmine/redmine-3.4.12.ebuild]: Could you sort this list alphabetically while at it :P

File: www-apps/redmine/redmine-3.4.12.ebuild [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+ruby_add_rdepend "
+	fastcgi? ( dev-ruby/fcgi )
```
Comment (juippis) [inline www-apps/redmine/redmine-3.4.12.ebuild]: This is looking better, just move the conditional block below unconditional ones.

File: www-apps/redmine/redmine-3.4.12.ebuild [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+
+REDMINE_DIR="/var/lib/${PN}"
+
+all_ruby_prepare() {
+	rm -r log files/delete.me .github || die
+
+	# bug #406605
+	rm .{git,hg}ignore || die
+
+	cat > "${T}/50${PN}" <<-EOF || die
```
Comment (juippis) [inline www-apps/redmine/redmine-3.4.12.ebuild]: You can skip using `${T}` here and just do this with `newenvd` in src_install, check 
https://gitweb.gentoo.org/repo/gentoo.git/tree/mail-client/thunderbird-bin/thunderbird-bin-68.2.0.ebuild#n151
for example.

File: www-apps/redmine/redmine-3.4.12.ebuild [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+		CONFIG_PROTECT="${EROOT%/}${REDMINE_DIR}/config"
+		CONFIG_PROTECT_MASK="${EROOT%/}${REDMINE_DIR}/config/locales ${EROOT%/}${REDMINE_DIR}/config/settings.yml"
+	EOF
+
+	# remove ldap staff module if disabled to avoid #413779
+	use ldap || rm app/models/auth_source_ldap.rb || die
```
Comment (juippis) [inline www-apps/redmine/redmine-3.4.12.ebuild]: This probably doesnt work as intended. Please open it as `if use ldap; then...`

File: www-apps/redmine/redmine-3.4.12.ebuild [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+		elog "# emerge --config \"=${CATEGORY}/${PF}\""
+		elog
+		elog "Installation notes are at official site"
+		elog "http://www.redmine.org/wiki/redmine/RedmineInstall"
+	fi
+}
+
+pkg_config() {
+	# Remove old lock file
+	rm -fr "${EROOT%/}${REDMINE_DIR}/Gemfile.lock"
```
Comment (juippis) [inline www-apps/redmine/redmine-3.4.12.ebuild]: Why `-r`, is this a directory? 

Pretty much all of these comments apply to 2nd ebuild as well.

File: www-apps/redmine/redmine-3.4.12.ebuild (hunk 6 above)
Comment (winterheart) [reply www-apps/redmine/redmine-3.4.12.ebuild]: Currently only ruby24 has all dependencies

File: www-apps/redmine/redmine-3.4.12.ebuild (hunk 9 above)
Comment (winterheart) [reply www-apps/redmine/redmine-3.4.12.ebuild]: newenvd not work in all_ruby_*() functions.

**Comment by winterheart** [comment]
@juippis mostly done

**Comment by winterheart** [comment]
@juippis ping

File: www-apps/redmine/redmine-4.0.5.ebuild [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+	rm -r log files/delete.me .github || die
+
+	# bug #406605
+	rm .{git,hg}ignore || die
+
+	# newenvd not working here
+	cat > "${T}/50${PN}" <<-EOF || die
+		CONFIG_PROTECT="${EROOT%/}${REDMINE_DIR}/config"
+		CONFIG_PROTECT_MASK="${EROOT%/}${REDMINE_DIR}/config/locales ${EROOT%/}${REDMINE_DIR}/config/settings.yml"
+	EOF
```
Comment (juippis) [inline www-apps/redmine/redmine-4.0.5.ebuild]: This should work in `all_ruby_install` because it's src_install function. Similar to fowners and fperms.

File: www-apps/redmine/redmine-4.0.5.ebuild [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+
+	# Fixing versions in Gemfile
+	eapply "${FILESDIR}/${P}_gemfile_versions.patch"
+
+	sed -i -e "/group :development do/,/end$/d" Gemfile || die
+	sed -i -e "/group :test do/,/end$/d" Gemfile || die
```
Comment (juippis) [inline www-apps/redmine/redmine-4.0.5.ebuild]: Please add a comment about what these seds do next time.
